[Feature request]: Support macOS Monterey's "Focus" modes

### Version

- [X] I'm using version 1.9.0

### Existing requests

- [X] I've checked Existing requests

### What operating system are you using?

macOS

### Operating System Version

macOS 12.1

### Problem description

macOS Monterey has the [Focus](https://support.apple.com/guide/mac-help/set-up-a-focus-to-stay-on-task-mchl613dc43f/mac) feature that allows the user to have multiple different Focus modes beside Do Not Disturb. So one might also want to not show breaks when in certain Focus modes.

### Possible solution

Instead of a checkbox option "Show breaks even in Do Not Disturb mode", allow setting a list of Focus mode (e.g., "Do not show breaks in these Focus modes:") where if the user is in one of those Focus modes, the app will not show breaks.

Though it does not seem to be an easy way to access the Focus mode. One could use [`INFocusStatusCenter`](https://developer.apple.com/documentation/sirikit/infocusstatuscenter) but that would require native code. And I could not find how we could use AppleScript to get the current Focus mode.

### Alternative solutions

On iOS, with Automation, one could run a Shortcut when entering or leaving a Focus mode. However, macOS's Shortcuts.app does not seem to [support](https://www.reddit.com/r/shortcuts/comments/qpb9u7/is_there_a_way_to_run_a_shortcut_on_the_mac_when/) that yet. If macOS's Shortcuts.app adds support for this, we could easily just have two Automations: one runs a Shortcut that runs the `stretchly pause` command when entering a Focus mode, and one runs a Shortcut that runs the `stretchly resume` command when leaving a Focus mode (and then we could add a document in the repo showing how to set this up).

### Additional information

Related to https://github.com/hovancik/stretchly/issues/969.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

@hovancik The path was correct. This should be the full log from the failed start:

<details>
  <summary>Log lines</summary>

```log
[2023-10-08 07:23:58.628] [info]  Stretchly: initializing...
[2023-10-08 07:23:58.631] [info]  Stretchly: loading preferences
[2023-10-08 07:23:58.636] [info]  Stretchly: starting Idle time monitoring
[2023-10-08 07:23:58.636] [info]  Stretchly: starting Do Not Disturb monitoring
[2023-10-08 07:23:58.671] [info]  Stretchly: loading default break ideas
[2023-10-08 07:23:58.692] [info]  Stretchly: loading default break ideas
[2023-10-08 07:23:59.647] [error] Error: Cannot find module '@meetfranz/electron-notification-state'
Require stack:
- /Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js
- /Applications/Stretchly.app/Contents/Resources/app.asar/breaksPlanner.js
- /Applications/Stretchly.app/Contents/Resources/app.asar/main.js
-
    at Module._resolveFilename (node:internal/modules/cjs/loader:1082:15)
    at n._resolveFilename (node:electron/js2c/browser_init:2:117469)
    at Module._load (node:internal/modules/cjs/loader:927:27)
    at f._load (node:electron/js2c/asar_bundle:2:13330)
    at Module.require (node:internal/modules/cjs/loader:1148:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at get _doNotDisturb [as _doNotDisturb] (/Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js:37:14)
    at Timeout._onTimeout (/Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js:45:33)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)
[2023-10-08 07:24:26.808] [error] TypeError: Object has been destroyed
    at Timeout.checkVersion [as _onTimeout] (/Applications/Stretchly.app/Contents/Resources/app.asar/main.js:680:16)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)
[2023-10-08 07:24:28.868] [error] Error: Cannot find module '@meetfranz/electron-notification-state'
Require stack:
- /Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js
- /Applications/Stretchly.app/Contents/Resources/app.asar/breaksPlanner.js
- /Applications/Stretchly.app/Contents/Resources/app.asar/main.js
-
    at Module._resolveFilename (node:internal/modules/cjs/loader:1082:15)
    at n._resolveFilename (node:electron/js2c/browser_init:2:117469)
    at Module._load (node:internal/modules/cjs/loader:927:27)
    at f._load (node:electron/js2c/asar_bundle:2:13330)
    at Module.require (node:internal/modules/cjs/loader:1148:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at get _doNotDisturb [as _doNotDisturb] (/Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js:37:14)
    at Timeout._onTimeout (/Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js:45:33)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)
[2023-10-08 07:24:31.280] [error] Error: Cannot find module '@meetfranz/electron-notification-state'
Require stack:
- /Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js
- /Applications/Stretchly.app/Contents/Resources/app.asar/breaksPlanner.js
- /Applications/Stretchly.app/Contents/Resources/app.asar/main.js
-
    at Module._resolveFilename (node:internal/modules/cjs/loader:1082:15)
    at n._resolveFilename (node:electron/js2c/browser_init:2:117469)
    at Module._load (node:internal/modules/cjs/loader:927:27)
    at f._load (node:electron/js2c/asar_bundle:2:13330)
    at Module.require (node:internal/modules/cjs/loader:1148:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at get _doNotDisturb [as _doNotDisturb] (/Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js:37:14)
    at Timeout._onTimeout (/Applications/Stretchly.app/Contents/Resources/app.asar/utils/dndManager.js:45:33)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)
```

</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - adds Vietnamese translations
 - reset breaks shortcut (advanced option)
 - skip to the next break shortcut (advanced option)
+- respect Do Not Disturb on Linux distros (KDE, XFCE, GNOME)
 
 ## Changed
 - updated many translations

--- a/app/package.json
+++ b/app/package.json
@@ -21,7 +21,7 @@
   "homepage": "https://hovancik.net/stretchly",
   "dependencies": {
     "@electron/remote": "^2.0.12",
-    "@meetfranz/electron-notification-state": "^1.0.0",
+    "dbus-final": "github:Jelmerro/dbus-final",
     "electron-log": "^4.4.8",
     "electron-store": "^8.1.0",
     "humanize-duration": "^3.30.0",
@@ -34,6 +34,6 @@
     "rtl-detect": "^1.0.4",
     "semver": "^7.5.4",
     "windows-focus-assist": "^1.3.0",
-    "windows-notification-state": "^2.0.0"
+    "windows-quiet-hours": "^2.0.0"
   }
 }

--- a/app/utils/dndManager.js
+++ b/app/utils/dndManager.js
@@ -17,6 +17,9 @@ class DndManager extends EventEmitter {
     this.monitorDnd = true
     this._checkDnd()
     log.info('Stretchly: starting Do Not Disturb monitoring')
+    if (process.platform === 'linux') {
+      log.info(`System: Your Desktop seems to be ${process.env.XDG_CURRENT_DESKTOP}`)
+    }
   }
 
   stop () {
@@ -27,22 +30,68 @@ class DndManager extends EventEmitter {
     log.info('Stretchly: stopping Do Not Disturb monitoring')
   }
 
-  get _doNotDisturb () {
-    if (this.monitorDnd) {
-      let focusAssist
-      try {
-        focusAssist = require('windows-focus-assist').getFocusAssist().value
-      } catch (e) { focusAssist = -1 } // getFocusAssist() throw an error if OS isn't windows
+  async _isDndEnabledLinux () {
+    const dbus = require('dbus-final')
+    const bus = dbus.sessionBus()
+    try {
+      const obj = await bus.getProxyObject('org.freedesktop.Notifications', '/org/freedesktop/Notifications')
+      const properties = obj.getInterface('org.freedesktop.DBus.Properties')
+      const dndEnabled = await properties.Get('org.freedesktop.Notifications', 'Inhibited')
+      if (await dndEnabled.value) {
+        return true
+      }
+    } catch (e) {
+      // KDE is not running
+    }
 
-      return require('@meetfranz/electron-notification-state').getDoNotDisturb() || (focusAssist !== -1 && focusAssist !== 0)
+    try {
+      const obj = await bus.getProxyObject('org.xfce.Xfconf', '/org/xfce/Xfconf')
+      const properties = obj.getInterface('org.xfce.Xfconf')
+      const dndEnabled = await properties.GetProperty('xfce4-notifyd', '/do-not-disturb')
+      if (await dndEnabled.value) {
+        return true
+      }
+    } catch (e) {
+      // XFCE is not running
+    }
+
+    try {
+      const util = require('node:util')
+      const exec = util.promisify(require('node:child_process').exec)
+      const { stdout } = await exec('gsettings get org.gnome.desktop.notifications show-banners')
+      if (stdout.replace(/[^0-9a-zA-Z]/g, '') === 'false') {
+        return true
+      }
+    } catch (e) {
+      // Gnome / gsettings is not running
+    }
+
+    return false
+  }
+
+  async _doNotDisturb () {
+    // TODO also check for session state? https://github.com/felixrieseberg/electron-notification-state/tree/master#session-state
+    if (this.monitorDnd) {
+      if (process.platform === 'win32') {
+        let wfa = 0
+        try {
+          wfa = require('windows-focus-assist').getFocusAssist().value
+        } catch (e) { wfa = -1 } // getFocusAssist() throw an error if OS isn't windows
+        const wqh = require('windows-quiet-hours').getIsQuietHours()
+        return wqh || (wfa !== -1 && wfa !== 0)
+      } else if (process.platform === 'darwin') {
+        return require('macos-notification-state').getDoNotDisturb()
+      } else if (process.platform === 'linux') {
+        return await this._isDndEnabledLinux()
+      }
     } else {
       return false
     }
   }
 
   _checkDnd () {
-    this.timer = setInterval(() => {
-      const doNotDisturb = this._doNotDisturb
+    this.timer = setInterval(async () => {
+      const doNotDisturb = await this._doNotDisturb()
       if (!this.isOnDnd && doNotDisturb) {
         this.isOnDnd = true
         this.emit('dndStarted')

--- a/package.json
+++ b/package.json
@@ -142,7 +142,7 @@
   },
   "dependencies": {
     "@electron/remote": "^2.0.12",
-    "@meetfranz/electron-notification-state": "^1.0.0",
+    "dbus-final": "github:Jelmerro/dbus-final",
     "electron-log": "^4.4.8",
     "electron-store": "^8.1.0",
     "humanize-duration": "^3.30.0",
@@ -155,6 +155,6 @@
     "rtl-detect": "^1.0.4",
     "semver": "^7.5.4",
     "windows-focus-assist": "^1.3.0",
-    "windows-notification-state": "^2.0.0"
+    "windows-quiet-hours": "^2.0.0"
   }
 }

